### PR TITLE
New runner to handle data from a directory

### PR DESCRIPTION
**Pull request by akhilaananthram**
MultiRunner expects the data to be in a directory. It reads the csv files in the directory and selects the same number of instances from each file to train on.  It uses the file name as the category label for training.  For testing, it either tests on a provided file or on the remaining instances.  It does not test on anything used for training.

There are also changes to readCSV to get the ID of a comment.  Because of this, there are changes to baseline_experiment so it will still work.

@BoltzmannBrain

### CODE REVIEW

File: fluent/experiments/multi_runner.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        # Randomly sampled, not repeated
+        trainIdx = random.sample(xrange(length), split)
+        testIdx = [i for i in xrange(length) if i not in trainIdx]
+      trainIdxs.append(trainIdx)
+      testIdxs.append(testIdx)
+
+      trainIdSet.update([self.patterns[label][i]["id"] for i in trainIdx])
+
+    if self.test:
+      testIdxs = [i for i, testInstance in enumerate(self.testPatterns) if testInstance["id"] not in trainIdSet]
```
Comment (BoltzmannBrain) [inline fluent/experiments/multi_runner.py]: wrap at 80 chars

File: fluent/utils/csv_helper.py [hunk 2]
```diff
@@ -49,13 +50,43 @@ def readCSV(csvFile, sampleIdx, numLabels):
...
+  @param dirPath            (str)          Path to the directory.
+  @param sampleIdx          (int)          Column number of the text samples.
+  @param numLabels          (int)          Number of columns of category labels.
+  @param modify             (bool)         Map the unix friendly category names
+                                           to the actual names. 0 -> /, _ -> " "
+  
+  @return samplesDict       (defaultdict)  Keys are CSV names, values are
+      OrderedDicts, where the keys/values are as specified in readCSV().
+  
+  @return randoSamplesDict  (defaultdict)  Copy of samplesDict, but the order of
```
Comment (BoltzmannBrain) [inline fluent/utils/csv_helper.py]: delete

File: fluent/experiments/multi_runner.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+      if self.testDict:
+        self.testSamples = [(texter.tokenize(data[0]), data[1], id)
+                            for id, data in self.testDict.iteritems()]
+
+
+  def setupData(self, preprocess=False, sampleIdx=2):
+    """
+    Get the data from CSV and preprocess if specified.
+    One index in labelIdx implies the model will train on a single
```
Comment (BoltzmannBrain) [inline fluent/experiments/multi_runner.py]: update docstring

File: fluent/experiments/multi_runner.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy
+import os
+import random
+
+from fluent.experiments.runner import Runner
+from fluent.utils.csv_helper import readCSV, readDir
+
```
Comment (BoltzmannBrain) [inline fluent/experiments/multi_runner.py]: remove blank line

File: fluent/experiments/experiment.py [hunk 5]
```diff
@@ -115,6 +133,9 @@ def run(args):
   parser.add_argument("--resultsDir",
                       default="results",
                       help="This will hold the experiment results.")
+  parser.add_argument("--textPreprocess",
+                      action="store_true",
```
Comment (BoltzmannBrain) [inline fluent/experiments/experiment.py]: add `default=False` to be explicit

File: fluent/experiments/baseline_experiment.py [hunk 6]
```diff
@@ -126,7 +126,9 @@ def computeExpectedAccuracy(predictedLabels, dataPath):
   Compute the accuracy of the models predictions against what we expect it to
   predict; considers only single classification.
   """
-  _, expectedLabels = readCSV(dataPath, 2, [3])
+  dataDict = readCSV(dataPath, 2, [3])
+  expectedLabels = [data[1] for id, data in dataDict.iteritems()]
```
Comment (BoltzmannBrain) [inline fluent/experiments/baseline_experiment.py]: let's use a different variable name than `id` so as not to confuse with the built-in method `id()`, and for all files in this PR

**Comment by BoltzmannBrain** [comment]
@akhilaananthram left some nitpicks, otherwise :+1:

**Comment by akhilaananthram** [comment]
@BoltzmannBrain: responded to comments. Do you have a desired behavior for the case scott brought up in the other PR for readDir, when two files in different sub-directories have the same name?

**Comment by BoltzmannBrain** [comment]
@akhilaananthram I don't see multiple files of the same name being an issue. `readDir()` reads one directory per call, and we're only operating on the returned `files` list from `os.walk()` so there's no issue with subdirectories. If a user wishes to run an experiment w/ multiple directories, that's logic they should handle in their experiment, not in a utility function.

**Pull request merged**